idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,851 @@
+    degrees_of_freedom,
     large_residuals_set,
     variables_near_bounds_set,
 )
-import idaes.core.util.scaling as iscale
+from idaes.core.util.scaling import list_badly_scaled_variables
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+MAX_STR_LENGTH = 84
+TAB = " " * 4
+
+
+def _var_in_block(var, block):
+    parent = var.parent_block()
+    while parent is not None:
+        if parent is block:
+            return True
+        parent = parent.parent_block()
+    return False
+
+
+class DiagnosticsToolbox:
+    """
+    The IDAES Model DiagnosticsToolbox.
+
+    To get started:
+
+      1. Create an instance of your model - this does not need to be initialized yet.
+      2. Fix variables until you have 0 degrees of freedom - many of these tools presume
+        a square model, and a square model should always be the foundation of any more
+        advanced model.
+      3. Create an instance of the DiagnosticsToolbox and provide the model to debug as
+        the model argument.
+      4. Call the report_structural_issues() method.
+
+    Model diagnostics is an iterative process and you will likely need to run these
+    tools multiple times to resolve all issues. After making a change to your model,
+    you should always start from the beginning again to ensure the change did not
+    introduce any new issues; i.e., always start from the report_structural_issues()
+    method.
+
+    Note that structural checks do not require the model to be initialized, thus users
+    should start with these. Numerical checks require at least a partial solution to the
+    model and should only be run once all structural issues have been resolved.
+
+    Report methods will print a summary containing three parts:
+
+    1. Warnings - these are critical issues that should be resolved before continuing.
+      For each warning, a method will be suggested in the Next Steps section to get
+      additional information.
+    2. Cautions - these are things that could be correct but could also be the source of
+      solver issues. Not all cautions need to be addressed, but users should investigate
+      each one to ensure that the behavior is correct and that they will not be the source
+      of difficulties later. Methods exist to provide more information on all cautions,
+      but these will not appear in the Next Steps section.
+    3. Next Steps - these are recommended methods to call from the DiagnosticsToolbox to
+      get further information on warnings. If no warnings are found, this will suggest
+      the next report method to call.
+
+    """
+
+    def __init__(self, model: Block):
+        if not isinstance(model, Block):
+            raise ValueError("model argument must be an instance of a Pyomo Block.")
+        self.model = model
+        # TODO: Work out how to manage and document these
+        self.residual_tolerance = 1e-5
+        self.zero_tolerance = 1e-6
```

Comment:
These should probably be arguments in the constructor, right?

Reply:
The intention was to do that eventually - I just wasn't sure of the best way to do it yet (separate arguments versus passing to a Config Block). Suggestions are welcome.

Reply:
I think a config block makes sense. That way it's pretty easy to also pass these arguments into e.g. the `report_structural_issues` method without having to make sure the call signatures stay in sync.